### DISCUSSION TITLE
[RFC FS-1115] Tracking the use of InlineIfLambda in FSharp.Core

### DISCUSSION THREAD

**Discussion by kerams** [fsharp/fslang-design#656] [Language and core library RFC discussions]
Discussion for https://github.com/fsharp/fslang-design/blob/main/RFCs/FS-1115-InlineIfLambda-in-FSharp-Core.md

> There are no known drawbacks of using this with the above functions

- What would be an example of a drawback? Does a lambda get inlined even when it's used multiple times, increasing assembly size?
- Does there have to be a clear benefit to inlining with every function, i.e. we don't want to sprinkle the attribute everywhere willy-nilly?

Of note: https://github.com/dotnet/fsharp/issues/12388

**Comment by kerams** [comment]
The Option module and its functions immediately comes to mind as a candidate.

**Reply by TheAngryByrd** [reply]
Option module, Value Option, Result module all would be easy candidates for me.

**Reply by kerams** [reply]
Indeed, those functions are so small that they could at least be inlined (a few already are).

**Reply by TheAngryByrd** [reply]
If anyone wants to see my exploration with Benchmarks and SharpLab output for inlining some of these functions: https://github.com/demystifyfp/FsToolkit.ErrorHandling/pull/166

**Comment by vzarytovskii** [comment]
One drawback which comes to mind is compilation/codegen speed, it _may_ slow it down significantly is some cases.

**Reply by kerams** [reply]
Codegen should be offset by the fact that closures no longer need to be emitted?

**Comment by dsyme** [comment]
> What would be an example of a drawback? Does a lambda get inlined even when it's used multiple times, increasing assembly size?


Yes, the RFC assumes single-use functions

**Comment by dsyme** [comment]
> Does there have to be a clear benefit to inlining with every function, i.e. we don't want to sprinkle the attribute everywhere willy-nilly?

Hard to tell.   I don't know of any specific downsides though.

**Comment by dsyme** [comment]
> Option module, Value Option, Result module all would be easy candidates for me.

Yes, all these should have it.
